src/lib.rs

```diff
@@ -218,105 +258,140 @@ fn encoded_size(bytes_len: usize, config: Config) -> usize {
-                std::ptr::write(output_ptr, charset[((input_chunk >> 58) & 0x3F) as usize]);
-                std::ptr::write(output_ptr.offset(1), charset[((input_chunk >> 52) & 0x3F) as usize]);
-                std::ptr::write(output_ptr.offset(2), charset[((input_chunk >> 46) & 0x3F) as usize]);
-                std::ptr::write(output_ptr.offset(3), charset[((input_chunk >> 40) & 0x3F) as usize]);
-                std::ptr::write(output_ptr.offset(4), charset[((input_chunk >> 34) & 0x3F) as usize]);
-                std::ptr::write(output_ptr.offset(5), charset[((input_chunk >> 28) & 0x3F) as usize]);
-                std::ptr::write(output_ptr.offset(6), charset[((input_chunk >> 22) & 0x3F) as usize]);
-                std::ptr::write(output_ptr.offset(7), charset[((input_chunk >> 16) & 0x3F) as usize]);
+                ptr::write(output_ptr, encode_table[((input_chunk >> 58) & 0x3F) as usize]);
+                ptr::write(output_ptr.offset(1), encode_table[((input_chunk >> 52) & 0x3F) as usize]);
+                ptr::write(output_ptr.offset(2), encode_table[((input_chunk >> 46) & 0x3F) as usize]);
+                ptr::write(output_ptr.offset(3), encode_table[((input_chunk >> 40) & 0x3F) as usize]);
+                ptr::write(output_ptr.offset(4), encode_table[((input_chunk >> 34) & 0x3F) as usize]);
+                ptr::write(output_ptr.offset(5), encode_table[((input_chunk >> 28) & 0x3F) as usize]);
+                ptr::write(output_ptr.offset(6), encode_table[((input_chunk >> 22) & 0x3F) as usize]);
+                ptr::write(output_ptr.offset(7), encode_table[((input_chunk >> 16) & 0x3F) as usize]);
                 output_ptr = output_ptr.offset(8);
             }
 
-            input_index += input_chunk_len;
-            fast_loop_output_buf_len += 8;
+            input_index += fast_chunk_len;
         }
     }
 
-    unsafe {
-        // expand len to include the bytes we just wrote
-        raw.set_len(fast_loop_output_buf_len);
-    }
+    // Encode the 0 to 7 bytes left after the fast loop.
 
-    // encode the 0 to 7 bytes left after the fast loop
-
-    let rem = input_bytes.len() % 3;
-    let start_of_rem = input_bytes.len() - rem;
+    let rem = input.len() % 3;
+    let start_of_rem = input.len() - rem;
+    let slow_chunk_len = 3;
 
     // start at the first index not handled by fast loop, which may be 0.
-    let mut leftover_index = input_index;
-
-    while leftover_index < start_of_rem {
-        raw.push(charset[(input_bytes[leftover_index] >> 2) as usize]);
-        raw.push(charset[((input_bytes[leftover_index] << 4 | input_bytes[leftover_index + 1] >> 4) & 0x3f) as usize]);
-        raw.push(charset[((input_bytes[leftover_index + 1] << 2 | input_bytes[leftover_index + 2] >> 6) & 0x3f) as usize]);
-        raw.push(charset[(input_bytes[leftover_index + 2] & 0x3f) as usize]);
 
-        leftover_index += 3;
+    while input_index < start_of_rem {
+        unsafe {
+            ptr::write(output_ptr,
```

Comment:
This is probably undetectably faster than `push()` but I switched it to use `ptr::write` like the fast loop because I felt it was easier to read and understand when all the writes use the same function, indexing, etc.

Also, now that we write into a slice, `push()` isn't available!